internal/component/statefulset/statefulset.go

```diff
@@ -278,23 +277,20 @@ func (r _resource) checkAndPatchStsPodLabelsOnMismatch(ctx component.OperatorCon
 }
 
 func (r _resource) areStatefulSetPodsUpdatedAndReady(ctx component.OperatorContext, sts *appsv1.StatefulSet) (bool, error) {
-	if sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
+	if err := r.client.Get(ctx, getObjectKey(sts.ObjectMeta), sts); err != nil {
+		return false, err
+	}
+	if sts.Status.ObservedGeneration < sts.Generation {
 		return false, nil
 	}
-	podList := &corev1.PodList{}
-	if err := r.client.List(ctx, podList, client.InNamespace(sts.Namespace), client.MatchingLabels(sts.Spec.Template.Labels)); err != nil {
-		return false, err
+	if sts.Status.UpdateRevision != sts.Status.CurrentRevision {
+		return false, nil
 	}
-	if len(podList.Items) != int(*sts.Spec.Replicas) {
+	if sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
```

Comment:
```suggestion
if *sts.Spec.Replicas!=nil && sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
```
safety check

Reply:
Even if you remove the replicas from the spec, it will be always defaulted to 1. You can try it out. It is a pointer but it can never be nil when it gets persisted in etcd.

Reply:
IMO, it's a good practice to check for `nil`. I'm leaving the decision to you.
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/rollout_status.go#L80

Reply:
k8s code also check for `nil`
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/rollout_status.go#L137

Reply:
Usage of pointer in K8S code base is not good. Even though it is not possible to set replicas to nil but since it is a pointer we can add a nil check.